### ISSUE TITLE
Issue on https TLS v1.2 with node v0.12

### ISSUE THREAD

**Issue by avin45h** [request/request#1555]
Hi guys, 
I am facing a bug in using request with node 0.12. 
Node version : v0.12.0
request version : 2.53.0

I am hitting a server with http protocol ,  the server responds back with a 301 redirect to https URL. The server is using TLS 1.2. This generates the following error : 

```javascript
_tls_common.js:108
        c.context.setKey(options.key);
                  ^
Error: error:0906D06C:PEM routines:PEM_read_bio:no start line
    at Error (native)
    at Object.createSecureContext (_tls_common.js:108:19)
    at Object.exports.connect (_tls_wrap.js:850:21)
    at Agent.createConnection (https.js:82:14)
    at Agent.createSocket (_http_agent.js:194:16)
    at Agent.addRequest (_http_agent.js:166:23)
    at new ClientRequest (_http_client.js:154:16)
    at Object.exports.request (http.js:49:10)
    at Object.exports.request (https.js:136:15)
    at Request.start (/home/avinash/projects/reinc_clone/frontend/node_modules/request/request.js:905:30)
    at Request.end (/home/avinash/projects/reinc_clone/frontend/node_modules/request/request.js:1636:10)
    at end (/home/avinash/projects/reinc_clone/frontend/node_modules/request/request.js:676:14)
    at Immediate._onImmediate (/home/avinash/projects/reinc_clone/frontend/node_modules/request/request.js:690:7)
    at processImmediate [as _immediateCallback] (timers.js:358:17)
```

I have tried hitting the server with https directly as well, and it returns with same result. For the above case, Following are the request options (logged at line no : 903 in request.js). 
/* I have modified original domain names and end points in the following snipplet */

```javascript
{ domain: null,
  _events: 
   { error: [Function],
     complete: [Function],
     pipe: [ [Function], [Function] ],
     end: [Function] },
  _maxListeners: undefined,
  uri: 
   { protocol: 'https:',
     slashes: true,
     auth: null,
     host: 'myServerAddr',
     port: 443,
     hostname: 'myServerAddr',
     hash: null,
     search: '?session[id]=foo',
     query: 'session[id]=foo',
     pathname: '/myEndPoint/',
     path: '/myEndPoint/?session[id]=foo',
     href: 'https://myServerAddr/myEndPoint/?session[id]=foo' },
  callback: [Function],
  headers: 
   { 'X-USER-DEVICE-TYPE': 'mobile',
     host: 'myServerAddr' },
  key: 'address',
  readable: true,
  writable: true,
  setHeader: [Function],
  hasHeader: [Function],
  getHeader: [Function],
  removeHeader: [Function],
  method: 'GET',
  localAddress: undefined,
  qsLib: { stringify: [Function], parse: [Function] },
  pool: {},
  dests: [],
  __isRequestRequest: true,
  _callback: [Function],
  proxy: null,
  tunnel: true,
  _redirectsFollowed: 1,
  maxRedirects: 10,
  allowRedirect: [Function],
  followRedirects: true,
  followAllRedirects: false,
  redirects: 
   [ { statusCode: 301,
       redirectUri: 'https://myServerAddr/myEndPoint/?session[id]=foo' } ],
  setHost: true,
  originalCookieHeader: undefined,
  _disableCookies: true,
  _jar: undefined,
  port: 443,
  host: 'myServerAddr',
  path: '/myEndPoint/?session[id]=foo',
  _auth: 
   { hasAuth: false,
     sentAuth: false,
     bearerToken: null,
     user: null,
     pass: null },
  httpModule: 
   { Server: { [Function: Server] super_: [Object] },
     createServer: [Function],
     globalAgent: 
      { domain: null,
        _events: [Object],
        _maxListeners: undefined,
        defaultPort: 443,
        protocol: 'https:',
        options: [Object],
        requests: {},
        sockets: {},
        freeSockets: {},
        keepAliveMsecs: 1000,
        keepAlive: false,
        maxSockets: Infinity,
        maxFreeSockets: 256 },
     Agent: { [Function: Agent] super_: [Object] },
     request: [Function],
     get: [Function] },
  agentClass: 
   { [Function: Agent]
     super_: { [Function: Agent] super_: [Object], defaultMaxSockets: Infinity } },
  href: 'https://myServerAddr/myEndPoint/?session[id]=foo',
  ntick: true,
  response: 
   { _readableState: 
      { objectMode: false,
        highWaterMark: 16384,
        buffer: [],
        length: 0,
        pipes: null,
        pipesCount: 0,
        flowing: true,
        ended: true,
        endEmitted: true,
        reading: false,
        sync: true,
        needReadable: false,
        emittedReadable: false,
        readableListening: false,
        defaultEncoding: 'utf8',
        ranOut: false,
        awaitDrain: 0,
        readingMore: false,
        decoder: null,
        encoding: null,
        resumeScheduled: false },
     readable: false,
     domain: null,
     _events: { end: [Object] },
     _maxListeners: undefined,
     socket: 
      { _connecting: false,
        _hadError: false,
        _handle: null,
        _host: 'myServerAddr',
        _readableState: [Object],
        readable: false,
        domain: null,
        _events: [Object],
        _maxListeners: 0,
        _writableState: [Object],
        writable: false,
        allowHalfOpen: false,
        destroyed: true,
        bytesRead: 1776,
        _bytesDispatched: 388,
        _pendingData: null,
        _pendingEncoding: '',
        parser: null,
        _httpMessage: [Object],
        read: [Function],
        _consuming: true,
        _idleNext: null,
        _idlePrev: null,
        _idleTimeout: -1 },
     connection: 
      { _connecting: false,
        _hadError: false,
        _handle: null,
        _host: 'myServerAddr',
        _readableState: [Object],
        readable: false,
        domain: null,
        _events: [Object],
        _maxListeners: 0,
        _writableState: [Object],
        writable: false,
        allowHalfOpen: false,
        destroyed: true,
        bytesRead: 1776,
        _bytesDispatched: 388,
        _pendingData: null,
        _pendingEncoding: '',
        parser: null,
        _httpMessage: [Object],
        read: [Function],
        _consuming: true,
        _idleNext: null,
        _idlePrev: null,
        _idleTimeout: -1 },
     httpVersionMajor: 1,
     httpVersionMinor: 1,
     httpVersion: '1.1',
     complete: true,
     headers: 
      { server: 'nginx',
        date: 'Mon, 27 Apr 2015 13:57:28 GMT',
        'content-type': 'application/json',
        'transfer-encoding': 'chunked',
        connection: 'close',
        'set-cookie': [Object],
        expires: 'Thu, 19 Nov 1981 08:52:00 GMT',
        'cache-control': 'no-store, no-cache, must-revalidate, post-check=0, pre-check=0',
        pragma: 'no-cache',
        location: 'https://myServerAddr/myEndPoint/?session[id]=foo',
        'x-server': 'integration1' },
     rawHeaders: 
      [ 'Server',
        'nginx',
        'Date',
        'Mon, 27 Apr 2015 13:57:28 GMT',
        'Content-Type',
        'application/json',
        'Transfer-Encoding',
        'chunked',
        'Connection',
        'close',
        'Set-Cookie',
        'j_udt=mobile; expires=Mon, 04-May-2015 13:57:25 GMT; path=/',
        'Set-Cookie',
        'PHPSESSID=foo; expires=Sat, 24-Oct-2015 13:57:25 GMT; path=/',
        'Expires',
        'Thu, 19 Nov 1981 08:52:00 GMT',
        'Cache-Control',
        'no-store, no-cache, must-revalidate, post-check=0, pre-check=0',
        'Pragma',
        'no-cache',
        'Set-Cookie',
        'split_test_stripe=Yes; expires=Mon, 11-May-2015 13:57:28 GMT; path=/',
        'Set-Cookie',
        'split_test_solr_search=newSearch_100; expires=Mon, 11-May-2015 13:57:28 GMT; path=/',
        'Set-Cookie',
        'split_test_tn=e_100; expires=Mon, 11-May-2015 13:57:28 GMT; path=/',
        'Set-Cookie',
        'split_test_ofc=nc_100; expires=Mon, 11-May-2015 13:57:28 GMT; path=/',
        'Set-Cookie',
        'split_test_sugg_ui=oldSuggestions_100; expires=Mon, 11-May-2015 13:57:28 GMT; path=/',
        'Set-Cookie',
        'split_test_sugg_bst=old_100; expires=Mon, 11-May-2015 13:57:28 GMT; path=/',
        'Set-Cookie',
        'split_test_sugg_prod=old_100; expires=Mon, 11-May-2015 13:57:28 GMT; path=/',
        'Set-Cookie',
        'split_test_sf=on_100; expires=Mon, 11-May-2015 13:57:28 GMT; path=/',
        'Set-Cookie',
        'split_test_sticky=new_50; expires=Mon, 11-May-2015 13:57:28 GMT; path=/',
        'Location',
        'https://myServerAddr/myEndPoint/?session[id]=foo',
        'X-Server',
        'integration1',
        'Set-Cookie',
        'SERVERIDN=AA; path=/' ],
     trailers: {},
     rawTrailers: [],
     _pendings: [],
     _pendingIndex: 0,
     upgrade: false,
     url: '',
     method: null,
     statusCode: 301,
     statusMessage: 'Moved Permanently',
     client: 
      { _connecting: false,
        _hadError: false,
        _handle: null,
        _host: 'myServerAddr',
        _readableState: [Object],
        readable: false,
        domain: null,
        _events: [Object],
        _maxListeners: 0,
        _writableState: [Object],
        writable: false,
        allowHalfOpen: false,
        destroyed: true,
        bytesRead: 1776,
        _bytesDispatched: 388,
        _pendingData: null,
        _pendingEncoding: '',
        parser: null,
        _httpMessage: [Object],
        read: [Function],
        _consuming: true,
        _idleNext: null,
        _idlePrev: null,
        _idleTimeout: -1 },
     _consuming: true,
     _dumped: false,
     req: 
      { domain: null,
        _events: [Object],
        _maxListeners: undefined,
        output: [],
        outputEncodings: [],
        outputCallbacks: [],
        writable: true,
        _last: true,
        chunkedEncoding: false,
        shouldKeepAlive: false,
        useChunkedEncodingByDefault: false,
        sendDate: false,
        _removedHeader: {},
        _hasBody: true,
        _trailer: '',
        finished: true,
        _hangupClose: false,
        _headerSent: true,
        socket: [Object],
        connection: [Object],
        _header: 'GET /myEndPoint/?session[id]=foo HTTP/1.1\r\nX-USER-DEVICE-TYPE: mobile\r\ncontent-type: application/json\r\nhost: myServerAddr\r\nConnection: close\r\n\r\n',
        _headers: [Object],
        _headerNames: [Object],
        agent: [Object],
        socketPath: undefined,
        method: 'GET',
        path: '/myEndPoint/?session[id]=foo',
        parser: null,
        res: [Circular] },
     request: 
      { domain: null,
        _events: [Object],
        _maxListeners: undefined,
        uri: [Object],
        callback: [Function],
        headers: [Object],
        key: 'address',
        readable: true,
        writable: true,
        setHeader: [Function],
        hasHeader: [Function],
        getHeader: [Function],
        removeHeader: [Function],
        method: 'GET',
        localAddress: undefined,
        qsLib: [Object],
        pool: {},
        dests: [],
        __isRequestRequest: true,
        _callback: [Function],
        proxy: null,
        tunnel: true,
        _redirectsFollowed: 1,
        maxRedirects: 10,
        allowRedirect: [Function],
        followRedirects: true,
        followAllRedirects: false,
        redirects: [Object],
        setHost: true,
        originalCookieHeader: undefined,
        _disableCookies: true,
        _jar: undefined,
        port: 443,
        host: 'myServerAddr',
        path: '/myEndPoint/?session[id]=foo',
        _auth: [Object],
        httpModule: [Object],
        agentClass: [Object],
        href: 'https://myServerAddr/myEndPoint/?session[id]=foo',
        ntick: true,
        response: [Circular],
        originalHost: 'myServerAddr',
        originalHostHeaderName: 'host',
        agent: [Object],
        _started: true },
     toJSON: [Function: responseToJSON],
     caseless: { dict: [Object] },
     read: [Function] },
  originalHost: 'myServerAddr',
  originalHostHeaderName: 'host',
  agent: 
   { domain: null,
     _events: { free: [Function] },
     _maxListeners: undefined,
     defaultPort: 443,
     protocol: 'https:',
     options: { path: null },
     requests: {},
     sockets: {},
     freeSockets: {},
     keepAliveMsecs: 1000,
     keepAlive: false,
     maxSockets: Infinity,
     maxFreeSockets: 256 },
  _started: true }
```

**Comment by lalitkapoor** [comment]
is this a server that you are running/control?

**Comment by avin45h** [comment]
Yes, But indirectly. I have the admin access to the server, however the
certificate was generated by the other guy. The https URL that I am trying
to hit opens up fine on browser.

**Comment by avin45h** [comment]
One more thing, I shifted my node version back to 0.10.33, request module
worked perfectly with that version, so I am pretty sure it's a version
issue.

**Comment by lalitkapoor** [comment]
I'm not as familiar with the changes for 0.12, going to ping @simov to help or to redirect to someone who can help faster.

**Comment by simov** [comment]
@avin45h most probably that's not an issue with request, although I can't tell you what exactly is. There are very good suggestion for the exact same error on [serverfault.com](http://serverfault.com/search?q=Error%3A+error%3A0906D06C%3APEM+routines%3APEM_read_bio%3Ano+start+line) It might be that v0.10 have more relaxed handling of certificate files or something similar that was fixed in 0.12

**Comment by avin45h** [comment]
@simov @lalitkapoor Found the issue, we were using request module in a wrong way. Alongwith options, form and method passed to the request module, we were passing something called "key" for our internal data seperation with multiple async calls. The name "key" was conflicting in this case, simply changing the name from "key" to something like "data_identifier" fixed the issue. For a better description, we were passing this to promisified request : 

```javascript
var settings = { url: 'http://foo/bar/hadouken/007',
  method: 'get',
  headers: { 'content-type': 'application/json' },
  key: 'shoryuken' }

request(settings); //promisified request with bluebird
```
This triggered exception

```javascript
_tls_common.js:108
        c.context.setKey(options.key);
```

Simply changing the key in settings to something else solved it. :)
